src/mqtt_client/AtlasMqttClient.cpp

```diff
@@ -0,0 +1,78 @@
+#include "AtlasMqttClient.h"
+#include "AtlasMqttClient_callback.h"
+#include "AtlasMqttClient_actionListener.h"
+#include "AtlasMqttLogger.h"
+#include "AtlasMqttException.h"
+#include <mqtt/async_client.h>
+
+namespace Atlas
+{
+AtlasMqttClient* AtlasMqttClient::instance = nullptr;
+
+AtlasMqttClient& AtlasMqttClient::getInstance()
+{
+    if (instance == nullptr)
+    {
+        instance = new AtlasMqttClient;
+    }
+    return *instance;
+}
+
+void AtlasMqttClient::destroyInstance()
+{
+    if (instance != nullptr)
+    {
+        delete instance;
+        instance = nullptr;
+    }
+}
+
+void AtlasMqttClient::publishMessage(const string address, const string clientID, const string topic, 
+                    const string message, int QoS)
+{
+    AtlasMqttLogger::getInstance().logMessage("Initializing client (" + clientID + 
+                        ") connection to server (" + address + ")");
+
+    mqtt::async_client client(address, clientID);
+    AtlasMqttClient_callback cb;
+    client.set_callback(cb);
+    mqtt::connect_options connOps;
+    
+    try
+    {
+        mqtt::token_ptr connTok = client.connect(connOps);
+        connTok->wait();
+        AtlasMqttLogger::getInstance().logMessage("Connection from " + clientID + " to " +
+                        address + " is establshed.");
+
+        mqtt::message_ptr pubMsg = mqtt::make_message(topic, message);
+        pubMsg->set_qos(QoS);        
+        mqtt::delivery_token_ptr pubTok;
+        AtlasMqttClient_actionListener listener;
+        pubTok = client.publish(pubMsg, nullptr, listener);
+        pubTok->wait();
+        AtlasMqttLogger::getInstance().logMessage("Message sent with QoS " + QoS);
+
+        auto tokens = client.get_pending_delivery_tokens();
+        if (!tokens.empty())        
+            throw "Error: Not all deliveries have finished. Pending delivery tokens exist.";        
+        
+        connTok = client.disconnect();
```

Comment:
Why disconnect here, can't we cache the connection?

Reply:
Modified the code. Disconnecting is now made by the user by calling a distinct method.